### PR TITLE
Support for metadata.generateName (CSA)

### LINKED ISSUE #2539: generate_name prefix seems to not be honored
### What happened?

I have both a `pv` and `pvc` being created, I specify the `pulumi_kubernetes.meta.v1.ObjectMetaArgs` `generate_name` attribute on each which should specify a prefix for the name.  But the name that results is not using my specified prefix, it appears to be using the `resource_name` of the object as the prefix instead.

### Expected Behavior

For example, my code for generating a `pvc` from the `pv` I created before
```
        # Create the PVC from the PV
        self.pvc = pulumi_kubernetes.core.v1.PersistentVolumeClaim(
            f'{self.cluster.name}-workbench-pvc',
            args=pulumi_kubernetes.core.v1.PersistentVolumeClaimInitArgs(
                metadata=pulumi_kubernetes.meta.v1.ObjectMetaArgs(
                    generate_name='workbench-pvc',
                    namespace='datascience'
                ),
                spec=pulumi_kubernetes.core.v1.PersistentVolumeClaimSpecArgs(
                    access_modes=['ReadWriteMany'],
                    resources=pulumi_kubernetes.core.v1.ResourceRequirementsArgs(
                        requests=capacity
                    ),
                    storage_class_name='',
                    volume_name=pv.metadata.name
                )
            ),
            opts=pulumi.ResourceOptions(provider=self.cluster.provider,
                                        parent=pv)
        )
```

I would expect the name of the `pvc` to be `workbench-pvc-<randomstring>` but instead I get `<cluster_name>-workbench-pvc-<randomstring>`

our cluster name is `internal-eks01` in this case
```
internal-eks01-workbench-pvc-c1aa4e6e         Bound    internal-eks01-workbench-pv-abae85d9         1200Gi     RWX                           55d
```

### Steps to reproduce

Create a `pvc` from a `pv` using the code below

 ```
        # Create the PVC from the PV
        self.pvc = pulumi_kubernetes.core.v1.PersistentVolumeClaim(
            f'{self.cluster.name}-workbench-pvc',
            args=pulumi_kubernetes.core.v1.PersistentVolumeClaimInitArgs(
                metadata=pulumi_kubernetes.meta.v1.ObjectMetaArgs(
                    generate_name='workbench-pvc',
                    namespace='datascience'
                ),
                spec=pulumi_kubernetes.core.v1.PersistentVolumeClaimSpecArgs(
                    access_modes=['ReadWriteMany'],
                    resources=pulumi_kubernetes.core.v1.ResourceRequirementsArgs(
                        requests=capacity
                    ),
                    storage_class_name='',
                    volume_name=pv.metadata.name
                )
            ),
            opts=pulumi.ResourceOptions(provider=self.cluster.provider,
                                        parent=pv)
        )
```

it will not use the `generate_name` prefix, it instead uses the `resource_name` as a prefix

### Output of `pulumi about`

```
❯ pulumi about
CLI
Version      3.77.1
Go Version   go1.20.7
Go Compiler  gc

Plugins
NAME                     VERSION
aws                      5.41.0
aws-native               0.72.0
kubernetes               3.28.1
kubernetes-cert-manager  0.0.5
postgresql               3.8.0
python                   unknown

Host
OS       darwin
Version  12.6.8
Arch     arm64

This project is written in python: executable='/Users/jforest/.pyenv/shims/python3' version='3.9.11
'

<SNIP SOME RESOURCES OUT>

Found no pending operations associated with jforest

Current Stack: jforest/aws-main-services/jforest
Backend
Name           pulumi.com
URL            https://app.pulumi.com/jforest
User           jforest
Organizations  jforest
```

### Additional context

versions of pulumi libraries:

```
❯ pip list | grep pulumi
pulumi                         3.68.0
pulumi-aws                     5.41.0
pulumi-aws-native              0.72.0
pulumi-kubernetes              3.28.1
pulumi-kubernetes-cert-manager 0.0.5
pulumi-postgresql              3.8.0
```

I would like to upgrade `pulumi-kubernetes` to 4.1.0, but `pulumi-kubernetes-cert-manager` will not allow versions greater than 3.x.  https://github.com/pulumi/pulumi-kubernetes-cert-manager/issues/21 and https://github.com/pulumi/pulumi-kubernetes-cert-manager/pull/22

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by EronWright**
### Proposed changes

Closes #2539

This PR implements support for `.metadata.generateName` in CSA mode. There's two major aspects to this:
1. Adjust the auto-naming logic to consider `.metadata.generateName` to be a variant of auto-naming. Pulumi will not assign an auto-name to a new resource that has `generateName`, and upon delete will use the replace-then-delete technique.
2. Use the output properties for await logic, since the inputs may or may not contain an object name. Note that auto-naming populates the inputs, whereas `generateName` doesn't.

This PR also fixes an autonaming bug where the old auto-name would erroneously be adopted, overwriting a computed name ([ref](https://github.com/pulumi/pulumi-kubernetes/pull/2594/files#diff-d3cf976a01c69136f5d733140f5e543802c3e326fb5ce16099dde403086411f4R50-R55)). It is in-scope because there's significant interplay between autonames and generateName.  (see https://github.com/pulumi/pulumi-kubernetes/pull/1618)

### Tests
A new integration test is provided to test the use of `metadata.generateName`.  It tests creation, update, replacement, and promotion from `.generateName` to `.name`.  ([ref](https://github.com/pulumi/pulumi-kubernetes/pull/2594/files#diff-7e85ce707283b1d281d971a8e5b0c49b959b0803ba3437dc9dbd422552326835R269-R400))

The existing autonaming test is enhanced to test how autonaming takes precedence over generateName, in the update case. This is to ensure backwards compatibility, e.g. in the edge case that an existing object has a `generateName` field (which Pulumi ignored until now). ([ref](https://github.com/pulumi/pulumi-kubernetes/pull/2594/files#diff-7e85ce707283b1d281d971a8e5b0c49b959b0803ba3437dc9dbd422552326835R220-R249))

### CODE REVIEW

**Comment by EronWright** [comment]
@lblackstone could you give me some feedback on this PR? Two aspects of note:
1. Do you agree that the "auto-import" thing that happens when you using SSA is actually a bug? In order to use `generateName`, one must use `create` rather than `patch`.
2. Why do the awaiters use the inputs and shouldn't they use the outputs? The inputs don't have the generated name, obviously.

**Comment by lblackstone** [comment]
> 1. Do you agree that the "auto-import" thing that happens when you using SSA is actually a bug? In order to use `generateName`, one must use `create` rather than `patch`. Pulumi should throw when an object already exists, unless you're using the Patch resource variant.

1. The "upsert" behavior was intentional with the idea that it's like running `kubectl apply`. If the cluster state already matches the declaration (desired state), then it shouldn't error. I think this difference was a side effect of using `Create` vs `Patch` previously. If we need to switch back to using `Create` for other reasons, we should see if it's possible to maintain this behavior where it succeeds if the state already matches the declaration. Alternatively, we could change the semantics to only upsert on Patch resources, but that is a break from the way the SSA behavior was designed and tested, so this would need to be done thoughtfully.


> 2. Why do the awaiters use the inputs and shouldn't they use the outputs? The inputs don't have the generated name, obviously.

2. Can you give me a specific example? From what I recall, all of the awaiters start watching event streams, so the input was used to figure out the name of the resource to watch. Computed outputs aren't available during preview, so that would be the most likely behavior change. It's possible that using outputs is more correct, but this is another change to take with care, since it could affect existing semantics.

File: provider/pkg/await/await.go [hunk 1]
```diff
@@ -162,6 +163,8 @@ func Creation(c CreateConfig) (*unstructured.Unstructured, error) {
 	// nolint
 	// https://github.com/kubernetes/kubernetes/blob/54889d581a35acf940d52a8a384cccaa0b597ddc/pkg/kubectl/cmd/apply/apply.go#L94
 
+	patchResource := kinds.PatchQualifiedTypes.Has(c.URN.QualifiedType().String())
```
Comment (lblackstone) [inline provider/pkg/await/await.go:166]: nit: I'd call this `isPatchResource` since patch can be either a noun or a verb.

File: provider/pkg/provider/provider.go [hunk 2]
```diff
@@ -1358,9 +1358,8 @@ func (k *kubeProvider) Check(ctx context.Context, req *pulumirpc.CheckRequest) (
 	// needs to be `DeleteBeforeReplace`'d. If the resource is marked `DeleteBeforeReplace`, then
 	// `Create` will allocate it a new name later.
 	if len(oldInputs.Object) > 0 {
-		// NOTE: If old inputs exist, they have a name, either provided by the user or filled in with a
+		// NOTE: If old inputs exist, they MAY have a name, either provided by the user or filled in with a
 		// previous run of `Check`.
-		contract.Assertf(oldInputs.GetName() != "", "expected object name to be nonempty: %v", oldInputs)
```
Comment (lblackstone) [inline provider/pkg/provider/provider.go:1363]: Should double-check to ensure that removing this assertion doesn't have implications elsewhere in the provider. The code was written with the assumption that the name was always present, although the actual update value didn't always match the preview value in the case of auto-naming, IIRC.

**Review by lblackstone — COMMENTED** [review]
I think these changes generally look ok, but with the caveats about possible behavior changes I mentioned earlier. The await code has been mostly stable for 3+ years, and is a core part of many resource updates, so this is an area to tread carefully. Particular testing care is needed for preview differences with computed inputs, and for resources that change between static names and the auto-named varieties.

**Comment by EronWright** [comment]
@lblackstone the rationale for using Create is that `generateName` doesn't work with apply. If one wrote a Pulumi program with `generateName`, I would argue that you don't expect to bind to an existing object (a would-be import).  Also, it seems to me incongruous that Pulumi would delete an object that it didn't create. Of course, the `Patch` variants neither create nor delete. 

The above was my rationale for the change, of course it needs further discussion.

File: provider/pkg/await/await.go [hunk 3]
```diff
@@ -90,17 +90,19 @@ type ReadConfig struct {
 
 type UpdateConfig struct {
 	ProviderConfig
-	Previous *unstructured.Unstructured
-	Inputs   *unstructured.Unstructured
-	Timeout  float64
-	Preview  bool
+	OldInputs  *unstructured.Unstructured
+	OldOutputs *unstructured.Unstructured
+	Inputs     *unstructured.Unstructured
```
Comment (lblackstone) [inline provider/pkg/await/await.go:95]: I'd suggest splitting the await logic change into a separate PR since it's not immediately apparent from the changelog or commit message that it's changing in a fundamental way.

Alternatively, you could structure this as two commits / changelog entries in the same PR, and then not squash merge them.

File: provider/pkg/provider/provider.go [hunk 4]
```diff
@@ -1386,6 +1385,14 @@ func (k *kubeProvider) Check(ctx context.Context, req *pulumirpc.CheckRequest) (
...
 		}
 	}
+	if metadata.IsGenerateName(newInputs, news) {
+		if k.serverSideApplyMode {
+			return nil, fmt.Errorf("the `.metadata.generateName` field is not supported in Server-Side Apply mode")
+		}
+		if k.yamlRenderMode {
+			return nil, fmt.Errorf("the `.metadata.generateName` field is not supported in YAML rendering mode")
+		}
+	}
```
Comment (lblackstone) [inline provider/pkg/provider/provider.go:1395]: Would it be better to ignore the generateName field in these cases rather than erroring? You could still print a warning, but that would be less disruptive.

File: tests/sdk/nodejs/generatename/step2/index.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+// Copyright 2016-2019, Pulumi Corporation.
```
Comment (lblackstone) [inline tests/sdk/nodejs/generatename/step2/index.ts:1]: nit: outdated copyright

File: provider/pkg/await/await.go (hunk 3 above)
Comment (rquitales) [reply provider/pkg/await/await.go:95]: This seems like a good proposal. While the actual logic to enable `generateName` is dependent on the await logic changes, the reverse isn't true. I'm more in favour of splitting these into 2 separate PRs, with the await logic being merged first.

File: provider/pkg/await/awaiters.go [hunk 6]
```diff
@@ -301,19 +301,19 @@ func untilAppsDeploymentDeleted(config deleteAwaitConfig) error {
 		specReplicas, _ := deploymentSpecReplicas(d)
 
 		return watcher.RetryableError(
-			fmt.Errorf("deployment %q still exists (%d / %d replicas exist)", config.currentInputs.GetName(),
+			fmt.Errorf("deployment %q still exists (%d / %d replicas exist)", d.GetName(),
```
Comment (rquitales) [inline provider/pkg/await/awaiters.go:304]: Should we introduce a new method for our config structs that returns the name of the deployment object? This refactor would make it less error-prone when deciding on the deployment name to use, and will make it easier in the future should we decide on a different strategy for the deployment name (say, reverting back to using inputs).

File: provider/pkg/metadata/naming.go [hunk 7]
```diff
@@ -42,14 +45,45 @@ func AssignNameIfAutonamable(randomSeed []byte, obj *unstructured.Unstructured,
...
+// Note that autonaming is preferred over generateName for backwards compatibility.
+func AdoptOldAutonameIfUnnamed(newObj, oldObj *unstructured.Unstructured, newObjMap resource.PropertyMap) {
+	if md, ok := newObjMap["metadata"].V.(resource.PropertyMap); ok {
+		// Check if the .metadata.name is set and is a computed value. If so, do not auto-name.
+		if name, ok := md["name"]; ok && name.IsComputed() {
+			return
+		}
+	}
 	if newObj.GetName() == "" && IsAutonamed(oldObj) {
+		contract.Assertf(oldObj.GetName() != "", "expected nonempty name for object: %s", oldObj)
```
Comment (rquitales) [inline provider/pkg/metadata/naming.go:57]: Incorrect formatting here? The unstructured object won't be nicely formatted into a string. Should the resource URN be printed here instead?

File: provider/pkg/provider/provider.go (hunk 4 above)
Comment (rquitales) [reply provider/pkg/provider/provider.go:1395]: Error-ing would indicate an issue with the user's code/manifest and that they should fix it - similar to conflicts with SSA. I'm on the stance of returning an error for the user to resolve, as ignoring this field will result in the name used for the object being different to the intent of the declared manifest.

**Review by rquitales — CHANGES_REQUESTED** [review]

**Comment by EronWright** [comment]
Closing this PR in favor of:
- https://github.com/pulumi/pulumi-kubernetes/pull/2790
- (TBD)
